shared/src/question-generators/graph-algorithms/dijkstra.ts

```diff
@@ -0,0 +1,180 @@
+    const correctOrder = runDijkstra(graph, startNode)
+
+    console.log("Expected Order of Node Removal:", correctOrder.join(", "))
+
+    const checkFormat: FreeTextFormatFunction = (answer: FreeTextAnswer) => {
+      const order = parseNodeOrder(answer.text)
+
+      for (const node of order) {
+        if (!graph.nodes.some((n) => n.label === node)) {
+          return {
+            valid: false,
+            message: t(translations, lang, "invalid_node", { n: node }),
+          }
+        }
+      }
+
+      return { valid: true, message: "" }
+    }
+
+    const feedback = ({ text }: { text: string }) => {
+      const answerOrder = parseNodeOrder(text)
+
+      if (answerOrder.length !== correctOrder.length) {
+        return {
+          correct: false,
+          correctAnswer: correctOrder.join(", "),
+          feedbackText: t(translations, lang, "feedback_node_count"),
+        }
+      }
+
+      for (let i = 0; i < answerOrder.length; i++) {
+        if (answerOrder[i] !== correctOrder[i]) {
+          return {
+            correct: false,
+            correctAnswer: correctOrder.join(", "),
+            feedbackText: t(translations, lang, "feedback_order", {
+              n: answerOrder[i],
+              i: (i + 1).toString(),
+            }),
+          }
+        }
+      }
+
+      return { correct: true }
+    }
+
+    const question: FreeTextQuestion = {
+      type: "FreeTextQuestion",
+      name: DijkstraAlgorithm.name(lang),
+      path: serializeGeneratorCall({ generator: DijkstraAlgorithm, lang, parameters, seed }),
+      text: `${t(translations, lang, "description")}\n\nStart node: **${startNode}**\n\n${graph.toMarkdown()}`,
+      prompt: t(translations, lang, "prompt"),
+      placeholder: "A, B, C, ...",
+      checkFormat,
+      feedback,
+      typingAid: graph.nodes.map((node) => ({
+        text: `$${node.label}$`,
+        input: `${node.label},`,
+        label: `Insert node ${node.label}`,
+      })),
+    }
+    return { question }
+  },
+}
+
+function runDijkstra(graph: Graph, startLabel: string): string[] {
+  const distances: Record<string, number> = {}
+  const visited: Set<string> = new Set()
+  const priorityQueue: [string, number][] = []
+
+  graph.nodes.forEach((node) => {
+    const label = node.label ?? ""
+    distances[label] = Infinity
+  })
+  distances[startLabel] = 0
+  priorityQueue.push([startLabel, 0])
+
+  const orderOfRemoval: string[] = []
+
+  while (priorityQueue.length > 0) {
+    // sort queue by distance and remove the node with the smallest distance
+    priorityQueue.sort((a, b) => a[1] - b[1])
+    const [currentNode, currentDist] = priorityQueue.shift() as [string, number]
+    if (visited.has(currentNode)) continue
+
+    visited.add(currentNode)
+    orderOfRemoval.push(currentNode)
+
+    // explore neighbors
+    const currentNodeIndex = graph.nodes.findIndex((n) => n.label === currentNode)
+    if (currentNodeIndex === -1) continue
```

Comment:
This should never be -1 or? Maybe add a comment or something?

Reply:
That was a remnant I overlooked from a now resolved complication (it was possible for nodes to not exist then).
Since that's not relevant anymore I went ahead and removed the line.